Use JSON Schema for event definitions

opl- opened the pull request:
# Introduction

As a new member I've been trying to find documentation of the infrastructure and since it appears that there is very little of that available, I decided to do something about it. I'm using this PR as a chance to learn about the projects behind ESA and in the process improve the documentation.

This PR is the first step in making the mq-events repo useful. It does so by declaring events in the JSON Schema format with some extensions. For more information see the updated README file.

This PR closes #3 and makes the `json-schema` branch obsolete.

# Specifics

## Names (and TypeScript)

This PR standarizes directory, file and event names. This was split into two commits in order to make it easier for git to recognize the history of individual files.

The ultimate goal is to make it possible to use these schemas as TypeScript definitions. Because of that, I believe names of the events and their definition files should be standardized. This is both for readability as well as to keep the names consistent between the file names and generated types. That being said, this was not strictly necessary as json-schema-to-typescript already handles dash separated names.

It's also worth nothing that [some issues in json-schema-to-typescript](https://github.com/bcherny/json-schema-to-typescript/issues/243) are making it impossible to fully utilize the possibilities of JSON Schema, specifically in the case of the `NodeCG/SCTimerChange` event. There might be other options available and I'm planning to do more research later.

## Routing keys

The routing key format is currently defined as a single string, but I also considered using some sort of a JSON Schema inspired object that would match all valid routing keys for a given event. This would allow referencing definitions (an example of where this would be useful is an event like `NodeCG/SCTimerChange`, where the `desc` prop values are duplicated in `$meta.routingKey`), but in the end I decided that the benefits are too small to give up the readability.

This field should be not only computer but also human readable, and using long JSON objects which also reference other places would likely make it more complicated to process in both cases.

The current format makes it relatively easy to generate channel bindings, which I'm hoping will prove valuable when generating TypeScript declarations, specifically because TypeScript doesn't allow for Regex validated string types.

## VHosts

EDIT: It has been decided that vHosts are a configuration detail and the prop was removed in cacf1f6f.

Most events previously defined the `vHost` property as `not.sure`, which is not a very useful piece of information.

From what I understand, the value of this property will most likely be one of `esa_prod`, `esa_local1` or `esa_local2`, where `esa_prod` is used by tools like the Tracker, UserTool and OmnibarModeration, while the "local" VHosts are used for events specific to stream 1 and stream 2 respectively, which are emitted by tools like FlagCarrier and BigButton.

For this reason I would consider reusing the format of `$meta.routingKey` for `$meta.vHost` (or possibly only the choice sequence), which would allow declaring full VHost strings without the need for external information to understand the used values.

I would appreciate your input on this.

# Review

I requested review from the respective creators of the events in this repository. Please, verify that your events are described correctly, especially paying attention to:

- New types of fields that were previously marked as `"$type": "id"`. In most cases those were changed to `"type": "string"`.
- `required` object properties.
- Nullability of types.
- Additional notes below, if any.

If you notice any issues, please correct them and push the changes to the PR's branch.

## Per project requests

### BigButton (@popeter45)

- I couldn't find the repository for BigButton that emits those events.
- The `ButtonPress/time` property might be possible to replace with a ref to the `timestampObject` definition.
- The `ButtonRGB` event declared the color components as "256 bit". I have to assume the resolution is actually 8 bit.
- `$meta.ownerTool` is declared as `SpeedControler` for some schemas. Is this significant or should they be updated to `BigButton`?
    - ~~TODO: Unsurprisingly, zoton confirmed that this is indeed completely irrelevant to Speedcontrol.~~ Done.
- `$meta.exchange` and `$meta.routingKey` were derived from `mq_topics.txt`.

### FlagCarrier (@BtbN)
- The `time` property in `TagScanned` schema was updated to reflect the format described in [TagScannedEvent](https://github.com/esamarathon/FlagCarrier/blob/master/FlagCarrierMini/TagScannedEvent.cs).
- Added `TagScanned/flagcarrier.group` prop.

### NodeCG (@zoton2)

- Added `OBSSceneChange/gameScene` prop.
    - Correct.
- The `SCActiveRunChanged/run` prop could use a proper schema.
    - TODO: zoton switched to using JSON Schema in his code for v2, so it might be possible to reuse those schemas.

### OmnibarModeration (@CBenni)

- I was unable to find the code emitting the events for this project.
- It seems like most messages there send objects in the format of a Twitch chat message. Should those be migrated to use a definition do they only include relevant props?
    - After a meeting with zoton we decided that only the relevant props (i.e. ones used in the code) should be declared there.
- `$meta.exchange` and `$meta.routingKey` were derived from `mq_topics.txt`.

### Tracker (@zoton2)

- I was unable to find the code emitting those events and instead used the GDQ Tracker models.
    - https://github.com/esamarathon/mq-repeater/blob/master/src/http.ts
- Format of `DonationFullyProcessed/time_received` is unclear.
    - It's ISO 8601.
- Values of `DonationFullyProcessed/comment_state` were derived from the tracker. Should they stay like that?
    - Values are correct, the notes in the description are accurate.
- `$meta.exchange` and `$meta.routingKey` were derived from `mq_topics.txt`.
    - They are correct.

### UserTool (@CBenni)

- (Luckily for me) I was unable to find the code emitting the `UserUpdated` event.
- The schema is missing everything mentioned in the general list.
- `$meta.exchange` and `$meta.routingKey` were derived from `mq_topics.txt`.

opl- committed:
Standardize file names

opl- committed:
Use JSON Schema for event definitions

opl- committed:
Remove mq_topics.txt

jofo8948 reviewed (commented):
It does no longer feel meaningful to include the cost as meta. Since specifying Vhost is sort of like specifying what environment or server it will be found at. The answer is always going to be "many". "esa_prod", "esa_test", "esa_dev" are just some examples going forward.

I suggest we remove the field since it is more if a part of the connection settings than the actual event.

opl- commented:
Personally I would argue that `$meta.vHost` could be a useful property as long as it's utilized properly. It could signify which environment the event originates from, which is both helpful as a hint for where to expect the event and for building testing tools (as suggested by zoton2).

Additionally I believe that production, staging and development should be completely separated, which includes the RabbitMQ instance. This eliminates the need for specifying which environment the program is running in and allows simply specifying values like `esa_cloud`, `esa_local1` and `esa_local2`. The reasoning here is that a badly written MQ integration being deployed in, say, staging could then starve the production instance of resources.

However, I didn't change the current values as that wouldn't match what is actually found currently in production and would be confusing. Changing the VHost would require changing configuration of components running in production and is not something I have the power to do.

jofo8948 commented:
My point is that it should not matter which VHost it is published to. That is a connection detail, much like which server and on what port to connect to. It can change from environment to environment.
Also, don't forget that esa_local1 and 2 are VHosts on a different server, with the some of the events appearing on both "cloud" and the local ones. If you want documentation of that, there is a far better way of documenting it with a proper tag, but calling the field VHost is not doing anyone any favors.

And there is not much point setting up multiple MQ clusters right now. If anything, we need additional nodes in our current cluster. And if we do split them, not having to update all our event definitions will make it so much easier.

Basic rule: If it goes in the app configuration, don't include it here.

opl- commented:
You know? You're right. If it can be changed in the config it shouldn't be in the event definitions. The idea behind including those was that this could be used to document where the events originate from, but it would make more sense for this to be in documentation explaining the environment and not here.

zoton2 commented on definitions/OmnibarModeration/NewScreenedTweet.json:
```diff
@@ -0,0 +1,43 @@
+			"properties": {
+				"name": {
+					"type": "string",
+					"description": "Name of the Twitter user who sent the tweet."
```
Is this the handle or the written name on Twitter (that can be changed)? I am very sure it's the handle.

zoton2 commented on definitions/OmnibarModeration/NewScreenedSub.json:
```diff
@@ -0,0 +1,46 @@
+							]
+						}
+					},
+					"additionalProperties": false
```
Twitch can send more so maybe set this to true/remove it?

zoton2 commented on definitions/OmnibarModeration/NewScreenedCheer.json:
```diff
@@ -0,0 +1,48 @@
+							"description": "The amount of bits cheered by the user."
+						}
+					},
+					"additionalProperties": false
```
Twitch can send more so maybe set this to true/remove it?

zoton2 commented on definitions/OmnibarModeration/NewScreenedSub.json:
```diff
@@ -0,0 +1,46 @@
+					"type": "string",
+					"description": "The message following the system message."
+				},
+				"tags": {
```
Add a note here about how these are tags supplied by Twitch.

zoton2 commented on definitions/OmnibarModeration/NewScreenedCrowdControl.json:
```diff
@@ -0,0 +1,27 @@
+			"required": [
+				"trailing"
+			],
+			"properties": {
```
Add tags here and add a note about how these are tags supplied by Twitch.

zoton2 commented on definitions/OmnibarModeration/NewScreenedCheer.json:
```diff
@@ -0,0 +1,48 @@
+					"type": "string",
+					"description": "The message text."
+				},
+				"tags": {
```
Add a note here about how these are tags supplied by Twitch.

zoton2 commented on definitions/NodeCG/SCTimerChanged.json:
```diff
@@ -0,0 +1,50 @@
+			]
+		},
+		"teamID": {
+			"type": "integer",
```
This is now a UUID string.

zoton2 commented on definitions/NodeCG/SCTimerChanged.json:
```diff
@@ -0,0 +1,50 @@
+		},
+		"timer": {
+			"type": "object",
+			"description": "Current copy of the timer object from nodecg-speedcontrol. See https://github.com/speedcontrol/nodecg-speedcontrol/blob/dev/types/timer.d.ts"
```
https://github.com/speedcontrol/nodecg-speedcontrol/blob/master/schemas/reused/Timer.json

zoton2 commented on definitions/NodeCG/OBSSceneChanged.json:
```diff
@@ -0,0 +1,43 @@
+		"routingKey": "<event_name>.obs.scene.<scene_name>.{end,start}{,.gamescene}",
+		"ownerTool": "NodeCG"
+	},
+	"required": [
```
Add "gameScene" as a required field.

zoton2 commented on definitions/NodeCG/SCActiveRunChanged.json:
```diff
@@ -0,0 +1,29 @@
+		},
+		"run": {
+			"type": "object",
+			"description": "Object of the run data from nodecg-speedcontrol. If this is not supplied, no run is set. See https://github.com/speedcontrol/nodecg-speedcontrol/blob/dev/types/RunData.d.ts"
```
https://github.com/speedcontrol/nodecg-speedcontrol/blob/master/schemas/reused/RunData.json

zoton2 commented on definitions/BigButton/ButtonIP.json:
```diff
@@ -1,11 +1,24 @@
+		"name": "ButtonIP",
+		"exchange": "bigbutton",
+		"routingKey": "<button_id>.ip_addr",
+		"ownerTool": "SpeedControler"
```
This should be changed to "BigButton", I don't think it's significant.

zoton2 commented on definitions/BigButton/ButtonPress.json:
```diff
@@ -1,15 +1,44 @@
+		"name": "ButtonPress",
+		"exchange": "bigbutton",
+		"routingKey": "<button_id>.pressed",
+		"ownerTool": "SpeedControler"
```
This should be changed to "BigButton", I don't think it's significant.

zoton2 commented on definitions/BigButton/ButtonPress.json:
```diff
@@ -1,15 +1,44 @@
+	],
+	"properties": {
+		"button_id": {
+			"type": "string",
```
This is apparently an integer

zoton2 commented:
I went through most of this stuff earlier with opl, besides what he's edited in and what I've written in the notes, there's a couple of things that came up that we need to ask:

- SCTimerChanged: As nodecg-speedcontrol v2 now uses UUIDv4 for all IDs instead of integers, this means that the routing key will have the whole thing in it, including the `-`s, and we're not sure if this is good/correct/valid. Mainly wanting to know the use case for the IDs being there, @BtbN probably knows.
- OBSSceneChanged: Similar to above, we were curious why the scene name is in the routing key. @BtbN again.

BtbN commented:
I now treat the ID as a random but unique string, which it effectively is. Something additional to sort the runs on would be nice, but not too important.
The scene name is in there to be able to only subscribe to/from changes to a certain scene. Without having to manually filter.

BtbN commented on schemas/event-schema.json:
```diff
@@ -0,0 +1,77 @@
+		"$meta": {
+			"type": "object",
+			"required": [
+				"vHost",
```
The vHost can probably be removed entirely. It's a configuration detail of the specific server/application in question, not of the event.

BtbN commented on schemas/event-schema.json:
```diff
@@ -0,0 +1,77 @@
+				},
+				"routingKey": {
+					"$ref": "#/definitions/nonEmptyString",
+					"description": "RabbitMQ routing key (topic) this event is emitted to"
```
Given how a lot of our routingKeys are pretty complex and use various forms of shell variable/expansion syntax are more than just a single routingKey, this should probably be expanded to reflect that.

Following how the definitions look today, they are using <var_name> for variable elements that are filled with data, and {a,b,c} syntax to define elements that have multiple possible values.
Additionally, it might be good to define how multiple independent keys might look. I'd say just comma-separate them.

zoton2 replied:
What the parts of the routing key mean is described in the README at least.

opl- replied:
I added a reference to the README file in the description of `$meta.routingKey`. I feel like describing the entire format in the description here would be a little unnecessary, so hopefully this will give people enough of a hint for how to use this. Technically a regex could be cooked up for this too, but I feel like that's a little overkill.

opl- committed:
Fix ownerTool of BigButton events

opl- committed:
Fix BigButton IDs should be integers

opl- committed:
Fix gameScene in NodeCG/OBSSceneChanged not required

opl- committed:
Add a UUID definition

opl- committed:
Change NodeCG/SCTimerChanged/teamID to be a UUID

opl- committed:
Document the tags prop in OmnibarModeration events better

opl- committed:
Clarify description of OmnibarModeration/NewScreenedTweet/user.name

opl- committed:
Remove $meta.vHost

It has been established in #6 that the vHost is entirely dependent on the configuration and not actually guaranteed by the implementation. For that reason it is being removed.

opl- commented:
I made some changes, mostly to address some of the things we agreed on about vHosts and the changes I discussed with zoton last week.

As I'm still unsure what the best way to refer to schemas from external repositories would be, for now I didn't include the schemas from Speedcontrol. The options I'm considering so far are:
- References to (tagged?) rawgithub URLs
    - Seems reasonable, as long as the tooling supports it. Credit to zoton for the suggestion.
- Blatant copy-pasting of the schemas
    - Causes unnecessary repetition, in my opinion.
- git submodules
    - Seems a little overkill to clone entire repos just to reference a few schemas in them.

opl- committed:
Add reference to the README file for the $meta.routerKey prop
